src/win/fs.c

```diff
@@ -1940,6 +1940,166 @@ INLINE static void fs__stat_prepare_path(WCHAR* pathw) {
   }
 }
 
+INLINE static int fs__stat_split_path(const WCHAR* filename, WCHAR** dir,
+    WCHAR** file) {
+  size_t len, i;
+  DWORD dir_len;
+
+  if (filename == NULL) {
+    if (dir != NULL)
+      *dir = NULL;
+    *file = NULL;
+    return 0;
+  }
+
+  len = wcslen(filename);
+  i = len;
+  while (i > 0 && filename[--i] != '\\' && filename[i] != '/');
+
+  if (i == 0) {
+    if (dir) {
+      dir_len = GetCurrentDirectoryW(0, NULL);
+      if (dir_len == 0) {
+        return -1;
+      }
+      *dir = (WCHAR*)uv__malloc(dir_len * sizeof(WCHAR));
+      if (!*dir) {
+        uv_fatal_error(ERROR_OUTOFMEMORY, "uv__malloc");
+      }
+
+      if (!GetCurrentDirectoryW(dir_len, *dir)) {
+        uv__free(*dir);
+        *dir = NULL;
+        return -1;
+      }
+    }
+
+    *file = _wcsdup(filename);
+  } else {
+    if (dir) {
+      *dir = (WCHAR*)uv__malloc((i + 2) * sizeof(WCHAR));
+      if (!*dir) {
+        uv_fatal_error(ERROR_OUTOFMEMORY, "uv__malloc");
+      }
+      wcsncpy(*dir, filename, i + 1);
+      (*dir)[i + 1] = L'\0';
+    }
+
+    *file = (WCHAR*)uv__malloc((len - i) * sizeof(WCHAR));
+    if (!*file) {
+      uv_fatal_error(ERROR_OUTOFMEMORY, "uv__malloc");
+    }
+    wcsncpy(*file, filename + i + 1, len - i - 1);
+    (*file)[len - i - 1] = L'\0';
+  }
+
+  return 0;
+}
+
+INLINE static int fs__stat_directory(WCHAR* path, uv_stat_t* statbuf,
+    int do_lstat) {
+
+  HANDLE handle;
+  FILE_STAT_BASIC_INFORMATION stat_info;
+  FILE_FULL_DIR_INFORMATION* pdir_info;
+  IO_STATUS_BLOCK ioStatusBlock;
+  NTSTATUS status;
+  WCHAR buf[32 * 1024];
+  WCHAR* path_dirpath = NULL;
+  WCHAR* path_filename = NULL;
+  BOOL ret_val = FALSE;
+  DWORD ret_error = 0;
+
+  if (fs__stat_split_path(path, &path_dirpath, &path_filename) != 0) {
+    ret_error = GetLastError();
+    goto cleanup;
+  }
+
+  /* Get directory handle */
+  handle = CreateFileW(path_dirpath,
+                       FILE_LIST_DIRECTORY,
+                       FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
+                       NULL,
+                       OPEN_EXISTING,
+                       FILE_FLAG_BACKUP_SEMANTICS,
+                       NULL);
+
+  uv__free(path_dirpath);
+  path_dirpath = NULL;
+
+  if (handle == INVALID_HANDLE_VALUE) {
+    ret_error = GetLastError();
+    goto cleanup;
+  }
+
+  do {
+    /* Get files in the directory */
+    status = pNtQueryDirectoryFile(handle,
+                                   NULL,
+                                   0,
+                                   0,
+                                   &ioStatusBlock,
+                                   buf,
+                                   sizeof(buf),
+                                   FileFullDirectoryInformation,
+                                   FALSE,
+                                   NULL,
```

Comment:
It might be much easier to pass the FileName argument here, and then just asserting there is only one?

Reply:
I tried your suggestion but got the same error as in the issue (UV__EACCES).